README

Some examples of valid urls:

Simple install instructions:
apt:3ddesktop

Some additonal constrains:
apt:2vcard?minver=1.0
apt:owl,newbiedoc,python-pykaraoke?dist=gutsy
apt:python-pykaraoke?dist=gutsy,owl,newbiedoc
apt:cdda2wav?section=multiverse
apt:java?section=multiverse?section=universe
apt:adobe-flashplugin?channel=partner

Some basic substutions are also supported:
apt:adobe-flashplugin?channel=$distro-partner
apt:linux-restricted-modules-$kernel'


debian/changelog
apturl (0.5.2+linuxmint13) vanessa; urgency=medium

  * bump - rebuild using XZ compression

 -- Clement Lefebvre <root@linuxmint.com>  Mon, 20 Jun 2022 13:36:09 +0200

apturl (0.5.2+linuxmint12) vanessa; urgency=medium

  * Depend on mint-common

 -- Clement Lefebvre <root@linuxmint.com>  Mon, 20 Jun 2022 10:55:18 +0200

apturl (0.5.2+linuxmint11) ulyana; urgency=medium

  * UI: Fix underlined label

 -- Clement Lefebvre <root@linuxmint.com>  Mon, 22 Jun 2020 15:28:53 +0100

apturl (0.5.2+linuxmint10) ulyana; urgency=medium

  [ gogogogi ]
  * Update hr.po (#5)

  [ Michael Webster ]
  * Fix dep resolution, improve initial query dialog (#4)

 -- Clement Lefebvre <root@linuxmint.com>  Sun, 21 Jun 2020 13:55:38 +0100

apturl (0.5.2+linuxmint9) ulyana; urgency=medium

  * Switch from synaptic to mintcommon-aptdaemon backend

 -- Clement Lefebvre <root@linuxmint.com>  Sat, 25 Apr 2020 13:53:44 +0100

apturl (0.5.2+linuxmint8) ulyana; urgency=medium

  * Fix refreshing cache

 -- Clement Lefebvre <root@linuxmint.com>  Sat, 25 Apr 2020 11:19:29 +0100

apturl (0.5.2+linuxmint7) tara; urgency=medium

  * Fix xapp dependency
  * Set dialog title

 -- Clement Lefebvre <root@linuxmint.com>  Wed, 30 May 2018 13:35:20 +0100

apturl (0.5.2+linuxmint6) tara; urgency=medium

  * UI: don't skip taskbar, set icon properly

 -- Clement Lefebvre <root@linuxmint.com>  Wed, 30 May 2018 11:58:20 +0100

apturl (0.5.2+linuxmint5) sylvia; urgency=medium

  [ monsta ]
  * drop aptdaemon deps - not needed anymore (#2)

 -- Clement Lefebvre <root@linuxmint.com>  Sat, 04 Nov 2017 13:58:08 +0000

apturl (0.5.2+linuxmint4) serena; urgency=medium

  [ gogogogi ]
  * Fix po.hr

 -- Clement Lefebvre <root@linuxmint.com>  Sun, 06 Nov 2016 13:24:57 +0000

apturl (0.5.2+linuxmint3) sarah; urgency=medium

  * Updated translations

 -- Clement Lefebvre <root@linuxmint.com>  Thu, 05 May 2016 18:42:53 +0100

apturl (0.5.2+linuxmint2) sarah; urgency=medium

  * Remove channel/section support and dependencies on Ubuntu components and webkit

 -- Clement Lefebvre <root@linuxmint.com>  Thu, 05 May 2016 16:54:28 +0100

apturl (0.5.2+linuxmint1) sarah; urgency=medium

  * Removed QT interface
  * Remove backends (will be reimplemented without dependencies on Ubuntu components)
  * Specify Gtk version in gi
  * Implemented synaptics installations

 -- Clement Lefebvre <root@linuxmint.com>  Thu, 05 May 2016 16:32:05 +0100

apturl (0.5.2ubuntu11) xenial; urgency=medium

  * Drop webkit1 support, we're switching to pure webkit2 now

 -- Iain Lane <iain@orangesquash.org.uk>  Wed, 13 Jan 2016 11:48:22 +0000

apturl (0.5.2ubuntu10) xenial; urgency=medium

  * debian/apturl.install:
    - don't install the gconf schemas, it's probably useful to no one
      nowadays (would only work for applications using the GNOME2 libgnome)
      and creates a depends on gconf which we want to drop from the iso

 -- Sebastien Bacher <seb128@ubuntu.com>  Wed, 16 Dec 2015 11:23:10 +0100

apturl (0.5.2ubuntu9) wily; urgency=medium

  * data/apturl-gtk.ui: set the dimension on the scrolledwindow rather
    than the textview so the description is showing with a non zero height 

 -- Sebastien Bacher <seb128@ubuntu.com>  Thu, 15 Oct 2015 17:56:46 +0100

apturl (0.5.2ubuntu8) wily; urgency=medium

  * Use webkit2 in addition to webkit1. (LP: #1469221)

 -- Iain Lane <iain@orangesquash.org.uk>  Tue, 30 Jun 2015 12:59:59 +0100

apturl (0.5.2ubuntu7) wily; urgency=medium

  * Don't include Debian version in the python module version (LP: #1465549)

 -- Zygmunt Krynicki <zygmunt.krynicki@canonical.com>  Fri, 26 Jun 2015 11:01:58 +0200

apturl (0.5.2ubuntu6) vivid; urgency=medium

  [ Tim Lunn ]
  * Don't use legacy icon names, these are no longer available in
    adwaita-icon-theme. (LP: #1437150) 

  [ Iain Lane ]
  * For dialog-question, go back to using the stock icon name as Humanity
    doesn't yet have an icon with the new name.

 -- Tim Lunn <tim@feathertop.org>  Thu, 02 Apr 2015 13:53:05 +0100

apturl (0.5.2ubuntu5) vivid; urgency=medium

  * Switch gnome-icon-theme to adwaita-icon-theme, which is its
    successor. (LP: #1414613)

 -- Iain Lane <iain@orangesquash.org.uk>  Mon, 26 Jan 2015 14:55:39 +0000

apturl (0.5.2ubuntu4) trusty; urgency=medium

  * Drop libgtk2-perl Recommends to Suggests.

 -- Dimitri John Ledkov <xnox@ubuntu.com>  Sun, 16 Mar 2014 02:56:58 +0000

apturl (0.5.2ubuntu3) trusty; urgency=low

  * AptUrl/AptUrl.py: don't use decode on a str (lp: #1050826)

 -- Sebastien Bacher <seb128@ubuntu.com>  Wed, 11 Dec 2013 11:18:25 +0100

apturl (0.5.2ubuntu2) saucy; urgency=low

  * adjust backend to reuse the bits from update-manager (LP: #1200775)

 -- Michael Vogt <michael.vogt@ubuntu.com>  Tue, 08 Oct 2013 20:14:22 +0200

apturl (0.5.2ubuntu1) raring; urgency=low

  * AptUrl/gtk/backend/__init__.py: Drop "action-done" signal definition;
    update-manager's GTK backend already defines it with different arguments,
    so this is ineffective and confusing.
  * AptUrl/gtk/GtkUI.py, _on_backend_done(): Add additional error signal
    arguments, as defined by update-manager's backend.
  * AptUrl/gtk/backend/InstallBackend{Aptdaemon,Synaptic}.py: Call action-done
    signal with error arguments. (LP: #1103024)
  * data/apturl-gtk.ui: Drop obsolete "has_separator" property. Fixes a
    warning.

 -- Martin Pitt <martin.pitt@ubuntu.com>  Tue, 09 Apr 2013 13:23:37 +0200

apturl (0.5.1ubuntu6) quantal; urgency=low

  * AptUrl/gtk/backend/InstallBackendAptdaemon.py: Work around
    incompatible Update Manager API change (as described in Launchpad
    bug #1026257).  LP: #1020980

 -- Barry Warsaw <barry@ubuntu.com>  Wed, 18 Jul 2012 13:38:02 -0400

apturl (0.5.1ubuntu5) quantal; urgency=low

  * debian/control: Move python3-aptdaemon dependency to apturl binary
    package (from apturl-common) since only the gtk backend needs it.

 -- Barry Warsaw <barry@ubuntu.com>  Fri, 15 Jun 2012 09:49:13 -0400

apturl (0.5.1ubuntu4) quantal; urgency=low


tests/test_helpers.py
import unittest

from AptUrl.Helpers import parse_pkg


class Candidate:
    def __init__(self, descr):
        self.description = descr
        self.homepage = None


class MockPkg:
    def __init__(self, descr):
        self.candidate = Candidate(descr)


class HelpersTest(unittest.TestCase):
    def test_parse_pkg(self):
        pkgobj = MockPkg("summary\ndescr\n")
        self.assertEqual(parse_pkg(pkgobj), ('summary','descr\n',None))
        pkgobj = MockPkg("summary only")
        self.assertEqual(parse_pkg(pkgobj), ('summary only','',None))


tests/test_apturl_parse.py
import unittest

from AptUrl import Parser


class UrlTests(unittest.TestCase):
    mapping = { 'kernel':"2.6.23-17-generic",
                'distro':"hardy" }

    testValues = ( ('apt:3ddesktop', [{'package':'3ddesktop',
                                      'schema': 'apt'}] ),
                   ('apt://3dchess', [{'package':'3dchess',
                                      'schema': 'apt'}] ),
                   ('apt+http://launchpad.net/~mvo/ppa?package=2vcard',
                    [{'package':'2vcard', 'schema': 'apt+http',
                     'repo_url': 'http://launchpad.net/~mvo/ppa',
                     'dist': '/' }] ),
                   ('apt+http://archive.canonical.com?package=acroread?dist=feisty?section=commercial',
                    [{'package':'acroread', 'schema': 'apt+http',
                     'repo_url': 'http://archive.canonical.com',
                     'dist': 'feisty', 'section': ['commercial']}] ),
                   ('apt+http://archive.canonical.com?package=acroread?keyfile=ppa-key',
                    [{'package':'acroread', 'schema': 'apt+http',
                     'repo_url': 'http://archive.canonical.com',
                     'keyfile': 'ppa-key' }]),
                   ('apt:2vcard?minver=1.0',
                    [{'package':'2vcard', 'schema': 'apt',
                     'minver': '1.0' }]),
                   ('apt:owl,newbiedoc,python-pykaraoke?dist=gutsy',
                    [{'package':'owl', 'schema':'apt'},
                    {'package':'newbiedoc', 'schema':'apt'},
                    {'package':'python-pykaraoke', 'schema':'apt', 'dist':'gutsy'}]),
                   ('apt:python-pykaraoke?dist=gutsy,owl,newbiedoc',
                    [{'package':'python-pykaraoke', 'schema':'apt', 'dist':'gutsy'},
                    {'package':'owl', 'schema':'apt'},
                    {'package':'newbiedoc', 'schema':'apt'}]),
                   ('apt:cdda2wav?section=multiverse',
                    [{'package':'cdda2wav', 'section':['multiverse']}]),
                   ('apt:java?section=multiverse?section=universe',
                    [{'package':'java', 'section':['multiverse','universe']}]),
                   ('apt:adobe-flashplugin?channel=partner',
                    [{'package':'adobe-flashplugin', 'channel': 'partner'}]),
                   ('apt:adobe-flashplugin?channel=$distro-partner',
                    [{'package':'adobe-flashplugin', 'channel': 'hardy-partner'}]),
                   ('apt:linux-restricted-modules-$kernel',
                    [{'package':'linux-restricted-modules-2.6.23-17-generic'}]),
                   ('apt:java?section=multiverse&section=universe',
                    [{'package':'java', 'section':['multiverse','universe']}]),
                   ('apt:java?section=multiverse&section=universe&minver=1.0',
                    [{'package':'java',
                      'section':['multiverse','universe'],
                      'minver': "1.0" }]),
                   ('apt:dpkg:i386', [{ 'package' : 'dpkg:i386',
                                      }]),
                 )

    def test_valid_parse(self):
        for v, e in self.testValues:
            #print("Testing: '%s'" % v)
            result_list = Parser.parse(v, self.mapping)
            for dictionary in e:
                for key in dictionary:
                    self.assertEqual(getattr(result_list[e.index(dictionary)],
                                             key), dictionary[key])

    def test_pkgname_too_long(self):
        self.assertRaises(
            Parser.InvalidUrlException, Parser.parse,
            "apt:" + 100 * "veryverylong")


AptUrl/AptUrl.py
# Copyright (c) 2007-2008 Canonical
#
# AUTHOR:
# Michael Vogt <mvo@ubuntu.com>
# With contributions by Siegfried-A. Gevatter <rainct@ubuntu.com>
#
# This file is part of AptUrl
#
# AptUrl is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as published
# by the Free Software Foundation; either version 2 of the License, or (at
# your option) any later version.
#
# AptUrl is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with AptUrl; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

import apt
import apt_pkg

from . import Parser
from . import Helpers
from .Helpers import _

from optparse import OptionParser

import os
import os.path

# adding new repositories is currently disabled because some people have
# security concerns about this feature
allow_new_repositories = False

# return codes
(RESULT_OK,
 RESULT_CANCELT,
 RESULT_ERROR,
 RESULT_BADARGS) = list(range(4))

class AptUrlController(object):

    def __init__(self, ui):
        self.ui = ui

    def openCache(self):
        try:
            self.cache = apt.Cache()
        except SystemError as strerr:
            if not '/etc/apt/sources.list' in str(strerr):
                raise
            self.ui.error(_("Invalid /etc/apt/sources.list file"), strerr)
            return False
        if self.cache._depcache.broken_count > 0:
            err_header = _("Software index is broken")
            err_body = _("This is a major failure of your software "
                         "management system. Please check for broken packages "
                         "with synaptic, check the file permissions and "
                         "correctness of the file '/etc/apt/sources.list' and "
                         "reload the software information with: "
                         "'sudo apt-get update' and 'sudo apt-get install -f'."
                         )
            self.ui.error(err_header, err_body)
            return False
        return True

    def parseArgs(self):
        parser = OptionParser()
        parser.add_option("-p", "--http-proxy", dest="http_proxy",
                          default=None, help="use http proxy")
        (options, args) = parser.parse_args()

        # eval and add proxy
        if options.http_proxy is not None:
            proxy = options.http_proxy
            if not ":" in proxy:
                proxy += ":3128"
            os.environ["http_proxy"] = "http://%s" % proxy

        # parse
        try:
            apturl_list = Parser.parse(args[0])
        except IndexError as e:
            self.ui.error(_("Need a url to continue, exiting"))
            return []
        except Parser.InvalidUrlException as e:
            self.ui.error(_("Invalid url: '%s' given, exiting") % e.url,
                          str(e))
            return []
        return (apturl_list)

    def verifyInstall(self, apturl):
        " verify that the install package actually is installed "
        # check if the package got actually installed
        self.openCache()
        pkg = self.cache[apturl.package]
        if (not pkg.is_installed or
            pkg._pkg.current_state != apt_pkg.CURSTATE_INSTALLED or
            self.cache._depcache.broken_count > 0):
            return False
        return True

    def main(self):
        # global return code
        ret = RESULT_OK
        ui = self.ui

        # parse arguments
        apturl_list = self.parseArgs()
        if not apturl_list:
            return RESULT_BADARGS

        # open cache
        if not self.openCache():
            return RESULT_ERROR

        # now go over the url list
        for apturl in apturl_list:
            # FIXME: move this code block into a func like
            #        evalAptUrl()

            if not apturl.schema in ("apt", "apt+http"):
                self.ui.error(_("Can not deal with protocol '%s' ")
                              % apturl.schema)
                continue

            if apturl.refresh is not None:
                ui.doUpdate()
                if not self.openCache():
                    return RESULT_ERROR

            # now check the package
            if apturl.package not in self.cache:
                try:
                    package_in_cache = bool(self.cache._cache[apturl.package])
                except KeyError:
                    package_in_cache = False
                if package_in_cache:
                    ui.error(_("Package '%s' is virtual.") % apturl.package)
                    continue
                else:
                    ui.error(_("Could not find package '%s'.")
                             % apturl.package)
                    continue

            if (self.cache[apturl.package].is_installed and
                apturl.minver is None):
                ui.message(_("Package '%s' is already installed")
                           % apturl.package)
                continue

            # ask the user
            pkg = self.cache[apturl.package]
            (sum, desc, homepage) = Helpers.parse_pkg(pkg)
            if not ui.askInstallPackage(apturl.package, sum, desc, homepage):
                ret = RESULT_CANCELT
                continue

            # try to install it
            try:
                self.cache[apturl.package].mark_install()
            except SystemError as e:
                ui.error(_("Can not install '%s' (%s) ") % (apturl.package, e))
                continue
            if apturl.minver is not None:
                verStr = self.cache[apturl.package].candidate.version
                if apt_pkg.version_compare(verStr, apturl.minver) < 1:
                    ui.error(_(
                        "Package '%s' requests minimal version '%s', but "
                        "only '%s' is available") % (apturl.package,
                                                     apturl.minver,
                                                     verStr))
                    continue

            changes = self.cache.get_changes()
            additional_pkgs = []

            for pkg in changes:
                if pkg.marked_install:
                    additional_pkgs.append(pkg.name)

            # install it
            ui.doInstall(apturl, extra_pkg_names=additional_pkgs)

            if not self.verifyInstall(apturl):
                ret = RESULT_ERROR

        # return values
        return ret


AptUrl/gtk/GtkUI.py
import gi
gi.require_version('Gtk', '3.0')
gi.require_version('XApp', '1.0')
from gi.repository import Gtk
from gi.repository import Gdk
from gi.repository import GObject
from gi.repository import XApp
GObject.threads_init()

import os
import sys
import apt_pkg
import subprocess
import tempfile

from AptUrl.UI import AbstractUI
from AptUrl import Helpers
from AptUrl.Helpers import _

import mintcommon.aptdaemon

APTURL_UI_FILE = os.environ.get(
    # Set this envar to use a test .ui file.
    'APTURL_UI_FILE',
    # System file to use if the envar is not set.
    '/usr/share/apturl/apturl-gtk.ui'
    )


class GtkUI(AbstractUI):
    def __init__(self):
        Gtk.init_check(sys.argv)
        # create empty dialog
        self.dia_xml = Gtk.Builder()
        self.dia_xml.set_translation_domain("apturl")
        self.dia_xml.add_from_file(APTURL_UI_FILE)
        self.dia = self.dia_xml.get_object('confirmation_dialog')
        self.dia.start_available = lambda: Gtk.main_quit()
        self.dia.start_error = lambda: Gtk.main_quit()
        self.dia.exit = lambda: Gtk.main_quit()
        self.dia.realize()
        self.require_update = False

    # generic dialogs
    def _get_dialog(self, dialog_type, summary, msg="",
                    buttons=Gtk.ButtonsType.CLOSE):
        " internal helper for dialog construction "
        d = Gtk.MessageDialog(parent=self.dia,
                              flags=Gtk.DialogFlags.MODAL,
                              type=dialog_type,
                              buttons=buttons)
        d.set_title("")
        d.set_markup("<big><b>%s</b></big>\n\n%s" % (summary, msg))
        XApp.set_window_icon_name(d, "package-x-generic")
        d.set_keep_above(True)
        d.realize()
        d.get_window().set_functions(Gdk.WMFunction.MOVE)
        return d

    def error(self, summary, msg=""):
        d = self._get_dialog(Gtk.MessageType.ERROR, summary, msg)
        d.run()
        d.destroy()
        return False

    def message(self, summary, msg="", title=""):
        d = self._get_dialog(Gtk.MessageType.INFO, summary, msg)
        d.set_title(title)
        d.run()
        d.destroy()
        return True

    def yesNoQuestion(self, summary, msg, title="", default='no'):
        d = self._get_dialog(Gtk.MessageType.QUESTION, summary, msg,
                             buttons=Gtk.ButtonsType.YES_NO)
        d.set_title(title)
        res = d.run()
        d.destroy()
        if res != Gtk.ResponseType.YES:
            return False
        return True

    def askInstallPackage(self, package, summary, description, homepage):
        # populate the dialog
        dia = self.dia
        dia_xml = self.dia_xml
        header = _("Install additional software?")
        body = _("Do you want to install package '%s'?") % package
        dia.set_title(package)
        header_label = dia_xml.get_object('header_label')
        header_label.set_markup("<b><big>%s</big></b>" % header)
        body_label = dia_xml.get_object('body_label')
        body_label.set_label(body)
        description_text_view = dia_xml.get_object('description_text_view')
        tbuf = Gtk.TextBuffer()
        desc = "%s\n\n%s" % (summary, Helpers.format_description(description))
        tbuf.set_text(desc)
        description_text_view.set_buffer(tbuf)
        XApp.set_window_icon_name(dia, "package-x-generic")

        # check if another package manager is already running
        # FIXME: just checking for the existance of the file is
        #        not sufficient, it need to be tested if it can
        #        be locked via apt_pkg.get_lock()
        #        - but that needs to run as root
        #        - a dbus helper might be the best answer here
        #args = (update_button_status, dia_xml.get_object("yes_button"),
        #    dia_xml.get_object("infolabel"))
        #args[0](*args[1:])
        #timer_id = GObject.timeout_add(750, *args )

        # show the dialog
        res = dia.run()
        #GObject.source_remove(timer_id)
        if res != Gtk.ResponseType.YES:
            dia.hide()
            return False

        return True

    # progress etc
    def doUpdate(self):
        self.require_update = True

    def doInstall(self, apturl, extra_pkg_names=[]):
        self.dia.hide()
        packages = []
        packages.append(apturl.package)
        packages += extra_pkg_names
        self.install_packages(packages)

    def install_packages(self, package_names):
        self.apt = mintcommon.aptdaemon.APT(None)
        self.package_names = package_names
        self.busy = True
        if self.require_update:
            self.apt.set_finished_callback(self.on_update_before_install_finished)
            self.apt.update_cache()
        else:
            self.on_update_before_install_finished()
        while self.busy:
            while Gtk.events_pending():
                Gtk.main_iteration()

    def on_update_before_install_finished(self, transaction=None, exit_state=None):
        self.apt.set_finished_callback(self.on_install_finished)
        self.apt.set_cancelled_callback(self.on_install_finished)
        self.apt.install_packages(self.package_names)

    def on_install_finished(self, transaction=None, exit_state=None):
        del self.package_names
        del self.apt
        self.busy = False
        self.dia.exit()

if __name__ == "__main__":
    ui = GtkUI()
    ui.error("foo","bar")


AptUrl/kde/KdeUI.py
# -*- coding: utf-8 -*-

from PyQt4.QtCore import *
from PyQt4.QtGui import *
from PyQt4.QtWebKit import QWebView
from PyQt4 import uic
from PyKDE4.kdeui import *
from PyKDE4.kdecore import *

import os
import apt_pkg
import subprocess

from AptUrl.UI import AbstractUI
from AptUrl.Helpers import _

APTURL_UI_FILE = os.environ.get(
    # Set this envar to use a test .ui file.
    'APTURL_UI_FILE',
    # System file to use if the envar is not set.
    '/usr/share/apturl/apturl-qt.ui'
    )


class AptUrlDialog(KDialog):
    def __init__(self,parent=None):
        KDialog.__init__(self,parent)
        self.setWindowIcon(KIcon("application-x-deb"))

    def slotButtonClicked(self,button):
        if button in (KDialog.Apply, KDialog.Ok, KDialog.Yes):
            KDialog.accept(self)
        else:
            KDialog.reject(self)


class KdeUI(AbstractUI):
    def __init__(self):
        self.dialog = AptUrlDialog()

        self.d = QWidget(self.dialog)
        self.dialog.setMainWidget(self.d)

        uic.loadUi(APTURL_UI_FILE, self.d)
        self.d.image_label.setPixmap(KIcon("application-x-deb").pixmap(64,64))

    # generic dialogs
    def _get_dialog(self, dialog_type, summary, msg="", buttons=KDialog.Close):
        self.dialog.setButtons(KDialog.ButtonCode(buttons))
        self.d.header_label.setText("<h2>" + summary + "</h2>")
        self.d.body_label.setText(msg)
        # TODO: title = empty
        # TODO: keep above maintain across events
        # TODO: d.set_markup("<big><b>%s</b></big>\n\n%s" % (summary, msg))

    def error(self, summary, msg=""):
        if msg != "":
            KMessageBox.detailedError(AptUrlDialog(), summary, msg)
        else:
            KMessageBox.error(AptUrlDialog(), summary)
        return False

    def message(self, summary, msg="", title=""):
        if msg != "":
            summary += "\n\n" + msg
        KMessageBox.information(AptUrlDialog(), summary, title)
        return True

    def yesNoQuestion(self, summary, msg, title="", default='no'):
        self._get_dialog('', summary, msg,
                             buttons=KDialog.Yes | KDialog.No)
        self.d.setWindowTitle(title)
        res = self.dialog.exec_()
        if res != KDialog.Accepted:
            return False
        return True

    # specific dialogs
    def askEnableChannel(self, channel, channel_info_html):
        summary = "<h2>" + _("Enable additional software channel") + "</h2>"
        msg = _("Do you want to enable the following "
                "software channel: '%s'?") % channel
        self._get_dialog('', summary, msg,
                             buttons=KDialog.Yes | KDialog.No)
        webview = QWebView()
        webview.setHtml(channel_info_html)
        webview.setFixedSize(400,200)
        self.d.vlayout_right.addWidget(webview)
        res = self.dialog.exec_()
        if res != KDialog.Accepted:
            return False
        return True

    def doEnableSection(self, sections):
        cmd = ["kdesudo",
               "--",
               "software-properties-kde",
               "--enable-component", "%s" % ' '.join(sections)]
        try:
            output = subprocess.check_output(
                cmd,
                stdout=subprocess.PIPE,
                universal_newlines=True)
        except (OSError, subprocess.CalledProcessError) as e:
            print("Execution failed: %s" % e, file=sys.stderr)
            return True
        # FIXME: Very ugly, but kdesudo doesn't return the correct exit states
        print(output)
        if not output.startswith("Enabled the "):
            return False
        return True


    def doEnableChannel(self, channelpath, channelkey):
        cmd = ["kdesudo",
               "--",
               "install", "--mode=644","--owner=0",channelpath,
               apt_pkg.Config.FindDir("Dir::Etc::sourceparts")]
        res = subprocess.call(cmd, universal_newlines=True)
        if res != 0:
            return False
        # install the key as well
        if os.path.exists(channelkey):
            cmd = ["kdesudo",
                   "--",
                   "apt-key", "add",channelkey]
            res = subprocess.call(cmd, universal_newlines=True)
            if res != 0:
                return False
        return True

    def askInstallPackage(self, package, summary, description, homepage):
        header = "<h2>" + _("Install additional software?") + "</h2>"
        body = _("Do you want to install package '%s'?") % package
        self.d.header_label.setText(header)
        self.d.body_label.setText(body)

        self.dialog.setButtons(KDialog.ButtonCode(KDialog.Yes | KDialog.No))

        res = self.dialog.exec_()
        if res != KDialog.Accepted:
            return False
        return True

    # progress etc
    def doUpdate(self):
        subprocess.check_call([
            'qapt-batch',
            '--attach', str(self.dialog.winId()),
            '--update'
            ],
            universal_newlines=True)

    def doInstall(self, apturl):
        subprocess.check_call([
            'qapt-batch',
            '--attach', str(self.dialog.winId()),
            '--install',
            apturl.package
            ],
            universal_newlines=True)


if __name__ == "__main__":
    ui = KdeUI()
    ui.error("foo","bar")

# kate: space-indent on; indent-width 4; mixedindent off; indent-mode python;


AptUrl/gtk/backend/InstallBackendSynaptic.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# (c) 2005-2007 Canonical, GPL

import apt_pkg
import os
import tempfile
from gettext import gettext as _

from gi.repository import GObject

from UpdateManager.backend import InstallBackend


class InstallBackendSynaptic(InstallBackend):
    """ Install backend based on synaptic """

    def _run_synaptic(self, action, opt, tempf):
        """Execute synaptic."""
        try:
            apt_pkg.pkgsystem_unlock()
        except SystemError:
            pass
        cmd = ["/usr/bin/gksu", 
               "--desktop", "/usr/share/applications/update-manager.desktop", 
               "--", "/usr/sbin/synaptic", "--hide-main-window",  
               "--non-interactive", "--parent-window-id",
               "%s" % self.window_main.window.xid ]
        cmd.extend(opt)
        flags = GObject.SPAWN_DO_NOT_REAP_CHILD
        (pid, stdin, stdout, stderr) = GObject.spawn_async(cmd, flags=flags)
        # Keep a reference to the data tuple passed to
        # GObject.child_watch_add to avoid attempts to destroy it without a
        # thread context: https://bugs.launchpad.net/bugs/724687
        self.child_data = (action, tempf)
        GObject.child_watch_add(pid, self._on_synaptic_exit, self.child_data)

    def _on_synaptic_exit(self, pid, condition, data):
        action, tempf = data
        if tempf:
            tempf.close()
        self.emit("action-done", action, True, os.WEXITSTATUS(condition) == 0, None, None)

    def update(self):
        opt = ["--update-at-startup"]
        tempf = None
        self._run_synaptic(self.UPDATE, opt, tempf)

    def commit(self, pkgs_install, pkgs_upgrade, close_on_done):
        # close when update was successful (its ok to use a Synaptic::
        # option here, it will not get auto-saved, because synaptic does
        # not save options in non-interactive mode)
        opt = []
        if close_on_done:
            opt.append("-o")
            opt.append("Synaptic::closeZvt=true")
        # custom progress strings
        opt.append("--progress-str")
        opt.append("%s" % _("Please wait, this can take some time."))
        opt.append("--finish-str")
        opt.append("%s" %  _("Update is complete"))
        tempf = tempfile.NamedTemporaryFile(mode="w+")
        for pkg_name in pkgs_install + pkgs_upgrade:
            tempf.write("%s\tinstall\n" % pkg_name)
        opt.append("--set-selections-file")
        opt.append("%s" % tempf.name)
        tempf.flush()
        self._run_synaptic(self.INSTALL, opt, tempf)


AptUrl/Parser.py
# Copyright (c) 2007-2008 Canonical
#
# AUTHOR:
# Michael Vogt <mvo@ubuntu.com>
# With contributions by Siegfried-A. Gevatter <rainct@ubuntu.com>
#
# This file is part of AptUrl
#
# AptUrl is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as published
# by the Free Software Foundation; either version 2 of the License, or (at
# your option) any later version.
#
# AptUrl is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with AptUrl; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

import os
import string
from string import Template
from .Helpers import get_dist
from .Helpers import _

class InvalidUrlException(Exception):
    def __init__(self, url, msg=""):
        self.url = url
        self.message = msg
    def __str__(self):
        return self.message

MAX_URL_LEN=255

# substituion mapping
apturl_substitution_mapping = {
    "distro" : get_dist(),
    "kernel" : os.uname()[2] 
}

# whitelist for the uri
whitelist = []
whitelist.extend(string.ascii_letters)
whitelist.extend(string.digits)
whitelist.extend(['_',':','?','/','+','.','~','=','<','>','-',',','$','&'])


class AptUrl(object):
    " a class that contains the parsed data from an apt url "
    def __init__(self):
        self.package = None
        self.schema = None
        self.minver = None
        self.refresh = None
        # for added repos
        self.keyfile = None
        self.repo_url = None
        self.dist = '/'     
        # for known sections 
        self.section = []
        # for known channels
        self.channel = None

def is_format_package_name(string):
    " return True if string would be an acceptable name for a Debian package "
    
    return (string.replace("+", "").replace("-", "").replace(".", "").replace(":", "").isalnum() 
        and string.islower() and string[0].isalnum() and len(string) > 1)

def do_apt_url_substitution(apt_url, mapping):
    " substitute known templates against the field package and channel "
    for field in ["package","channel"]:
        if getattr(apt_url, field):
            s=Template(getattr(apt_url, field))
            setattr(apt_url, field, s.substitute(mapping))

def match_against_whitelist(raw_url):
    " test if the url matches the internal whitelist "
    for char in raw_url:
        if not char in whitelist:
            raise InvalidUrlException(
                raw_url, _("Non whitelist char in the uri"))
    return True

def set_value(apt_url, s):
    " set a key,value pair from string s to AptUrl object "
    (key, value) = s.split("=")
    try:
        if ' ' in value:
            raise InvalidUrlException(apt_url, _("Whitespace in key=value"))
        if type(getattr(apt_url, key)) == type([]):
            getattr(apt_url, key).append(value)
        else:
            setattr(apt_url, key, value)
    except Exception as e:
        raise InvalidUrlException(apt_url, _("Exception '%s'") % e)


def parse(full_url, mapping=apturl_substitution_mapping):
    " parse an apt url and return a list of AptUrl objects "
    # apt:pkg1?k11=v11?k12=v12,pkg2?k21=v21?k22=v22,...
    res = []
    
    if len(full_url) > MAX_URL_LEN:
        url = "%s ..." % full_url[0:(MAX_URL_LEN // 10)]
        raise InvalidUrlException(url, _("Url string '%s' too long") % url)

    # check against whitelist
    match_against_whitelist(full_url)
    for url in full_url.split(";"):
        if not ":" in url:
            raise InvalidUrlException(url, _("No ':' in the uri"))

        # now parse it

        (schema, packages) = url.split(":", 1)
        packages = packages.split(",")

        for package in packages:
            apt_url = AptUrl()
            apt_url.schema = schema
            # check for schemas of the form: apt+http://
            if schema.startswith("apt+"):
                apt_url.repo_url = schema[len("apt+"):] + ":" + package.split("?",1)[0]
            else:
                if "?" in package:
                    apt_url.package = package.split("?")[0].lstrip("/")
                else:
                    apt_url.package = package.lstrip("/")

            # now parse the ?... bits
            if "?" in package:
                key_value_pairs = package.split("?")[1:]
                for s in key_value_pairs:
                    if "&" in s:
                        and_key_value_pairs = s.split("&")
                        for s in and_key_value_pairs:
                            set_value(apt_url, s)
                    else:
                        set_value(apt_url, s)

            # do substitution (if needed) 
            do_apt_url_substitution(apt_url, mapping)
            
            # check if the package name is valid
            if not is_format_package_name(apt_url.package):
                raise InvalidUrlException(url, "Invalid package name '%s'" % apt_url.package)
            
            res.append(apt_url)
    return res    
